Improve DistributedCacheHelper. Closes #896

milleruntime opened the pull request:
Remove duplicate DistributedCacheHelper and made new hadoop use the old impl.
Pulled code from deprecated hadoop classes and made small improvements.

milleruntime commented:
> I don't know enough about setting cache files in the job configuration this way. It seems reasonable, but has it been tested?

I am looking to see if any of the MR tests use it.  I think I am going to "un" fully qualify some of the references to help finding its use.

ctubbsii commented:
The TokenFileIT should use it.

milleruntime commented:
> The TokenFileIT should use it.

That was a good one, it uses 2 of the 3.  RangePartitioner looks to be the only class that uses getLocalCacheFiles but our test doesn't cover that code.  I can add at Test for it.

ctubbsii commented:
I don't know enough about the `getCacheFiles()` and `getLocalCacheFiles()`, but my expectation would be that we always want to try to get the local ones, because those are the ones in the distributed cache... whereas the other ones aren't actually cached... they're just references to the files outside the cache (in main DFS).

milleruntime commented:
> I don't know enough about the `getCacheFiles()` and `getLocalCacheFiles()`, but my expectation would be that we always want to try to get the local ones, because those are the ones in the distributed cache... whereas the other ones aren't actually cached... they're just references to the files outside the cache (in main DFS).

Its very confusing, I am trying to decipher.  [Javadoc for addCacheFile](https://github.com/apache/hadoop/blob/34a931c0fa5853fbcaadf0211d6bbee881715748/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/filecache/DistributedCache.java#L286) and [getCacheFiles](https://github.com/apache/hadoop/blob/34a931c0fa5853fbcaadf0211d6bbee881715748/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/filecache/DistributedCache.java#L181).  The one for getLocalCacheFiles looks to be the most promising: 
<pre>
/**
   * Return the path array of the localized files.  Intended to be used
   * by user code.
   * @param conf Configuration that contains the localized files
   * @return A path array of localized files
   * @throws IOException
   * @deprecated Use {@link JobContext#getLocalCacheFiles()} instead
   * @see JobContext#getLocalCacheFiles()
   */
</pre>

So it looks like getCacheFiles() was not intended to be for users, maybe that would explain deprecation.

Also, maybe we should consider not supporting the older mapred code since this is one case where you have to use the mapreduce methods if you don't want to use deprecated.

milleruntime commented:
What is bizarre is...
addCacheFile -> uses "mapreduce.job.cache.files"
getCacheFiles -> uses "mapreduce.job.cache.files"
getLocalCacheFiles -> uses "mapreduce.job.cache.local.files"

Some internal mechanisms set the files in "mapreduce.job.cache.local.files"...

ctubbsii commented:
> Also, maybe we should consider not supporting the older mapred code since this is one case where you have to use the mapreduce methods if you don't want to use deprecated.

This would make it hard to support the file-based token passing. We could just pull the file directly from HDFS, in that case. However, I'm not sure it's true that it requires the new API. I thought the old `JobConf` class does extend the internal `JobContext` which has the method for retrieving the local cache files.

> Some internal mechanisms set the files in "mapreduce.job.cache.local.files"...

That's what I'd expect, but it's not clear to me the expected behavior that would explain how to use this properly.

milleruntime commented:
> > Also, maybe we should consider not supporting the older mapred code since this is one case where you have to use the mapreduce methods if you don't want to use deprecated.
> 
> This would make it hard to support the file-based token passing. We could just pull the file directly from HDFS, in that case. However, I'm not sure it's true that it requires the new API. I thought the old `JobConf` class does extend the internal `JobContext` which has the method for retrieving the local cache files.

AFIAK there are two different ways to use the cache:
- For mapred: `DistributedCache.addCacheFile()`.  The older `JobConf` object extends `Configuration`.
- For mapreduce: `Job.getInstance().addCacheFile()`  The newer `Job` class implements `org.apache.hadoop.mapreduce.JobContext`

And this is the best documentation I have found: https://stackoverflow.com/questions/21239722/hadoop-distributedcache-is-deprecated-what-is-the-preferred-api

milleruntime commented:
OK I reverted the impl to what it was and just made the new Partitioner call the old impl code.  Removed the deprecation since its now used by the new package.

@ctubbsii what do you think of these changes?  This is all I really wanted to accomplish with this PR.  Other ideas you have we can do as follow on.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/client/mapreduce/lib/partition/RangePartitioner.java:
```diff
@@ -131,8 +131,7 @@ public void setConf(Configuration conf) {
     URI uri = new Path(file).toUri();
-    org.apache.accumulo.core.clientImpl.mapreduce.lib.DistributedCacheHelper.addCacheFile(uri,
-        job.getConfiguration());
+    DistributedCacheHelper.addCacheFile(uri, job.getConfiguration());
```
Can we just call `job.addCacheFile()` or whatever here?

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/client/mapreduce/lib/partition/RangePartitioner.java:
```diff
@@ -91,8 +92,7 @@ private synchronized int getNumSubBins() {
       String cutFileName = conf.get(CUTFILE_KEY);
-      Path[] cf = org.apache.accumulo.core.clientImpl.mapreduce.lib.DistributedCacheHelper
-          .getLocalCacheFiles(conf);
+      Path[] cf = DistributedCacheHelper.getLocalCacheFiles(conf);
```
Is `job.getLocalCacheFiles()` deprecated?

milleruntime replied:
No that one is good.  I guess its safe to call Job.getInstance() here?

ctubbsii replied:
> I guess its safe to call Job.getInstance() here?

I'm not sure what that does. Can't we just pass the job context into this private method?

milleruntime replied:
We could but we would be doing the same thing in the method calling it.  It is called by the overridden getPartition() in Partitioner.  The only params from it are: `Text key, Writable value, int numPartitions`

ctubbsii replied:
Oh, that sucks... the partitioner has no native concept of job context. However, the job's configuration is passed in by the job context when it is initialized, because it implements Configurable. So, we could call `job.getInstance(getConf(), "dummy")` if it doesn't cause too much side-effects, so we can access the job's convenience method to get the local cache files. If the implementation is trivial enough, or if instantiating a new dummy job has unwanted side-effects, we could probably also just parse the config item storing the paths to the local cache files ourselves.

milleruntime replied:
I am not sure... it looks like any call "Creates a new job" https://hadoop.apache.org/docs/r2.4.1/api/org/apache/hadoop/mapreduce/Job.html#getInstance()

ctubbsii replied:
Yeah, but if it's just a container object that holds the config and provides methods, but doesn't actually do anything until you submit it, it's probably fine. We'd have to check the Hadoop 3 impl of Job.
